subprojects/core/src/main/java/org/gradle/process/internal/DefaultExecAction.java

```diff
@@ -16,27 +16,221 @@
 
 package org.gradle.process.internal;
 
-import org.gradle.initialization.BuildCancellationToken;
-import org.gradle.internal.file.PathToFileResolver;
+import org.gradle.process.BaseExecSpec;
+import org.gradle.process.CommandLineArgumentProvider;
 import org.gradle.process.ExecResult;
+import org.gradle.process.ProcessForkOptions;
 
-import java.util.concurrent.Executor;
+import java.io.File;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Map;
 
 /**
  * Use {@link ExecActionFactory} (for core code) or {@link org.gradle.process.ExecOperations} (for plugin code) instead.
  */
-public class DefaultExecAction extends DefaultExecHandleBuilder implements ExecAction {
-    public DefaultExecAction(PathToFileResolver fileResolver, Executor executor, BuildCancellationToken buildCancellationToken) {
-        super(fileResolver, executor, buildCancellationToken);
+public class DefaultExecAction implements ExecAction {
+
+    private final ClientExecHandleBuilder execHandleBuilder;
+    private boolean ignoreExitValue;
+
+    public DefaultExecAction(ClientExecHandleBuilder execHandleBuilder) {
+        this.execHandleBuilder = execHandleBuilder;
     }
 
     @Override
     public ExecResult execute() {
-        ExecHandle execHandle = build();
+        ExecHandle execHandle = execHandleBuilder.build();
         ExecResult execResult = execHandle.start().waitForFinish();
         if (!isIgnoreExitValue()) {
             execResult.assertNormalExitValue();
         }
         return execResult;
     }
+
+    @Override
+    public String getExecutable() {
```

Comment:
💭 We can probably simplify this class when we migrate to Provider API and remove all the setters, maybe also getters if they can be generated by Gradle. We can then setup `execHandleBuilder` in `execute()`

Reply:
❓ Of this context, what would make sense to capture as a `// TODO` here?